pkg/util/archive/zip.go

```diff
@@ -0,0 +1,261 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package archive provides functions to archive and unarchive files.
+package archive
+
+import (
+	"archive/zip"
+	"compress/flate"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+// Zip is an adapted implementation of (*Zip).Archive from
+// https://github.com/mholt/archiver/blob/v3.5.1/zip.go#L140
+// Copyright (c) 2016 Matthew Holt
+func Zip(sources []string, destination string) error {
+	if !strings.HasSuffix(destination, ".zip") {
+		return fmt.Errorf("%s must have a .zip extension", destination)
+	}
+	if fileExists(destination) {
+		return fmt.Errorf("file already exists: %s", destination)
+	}
+	dir := filepath.Dir(destination)
+	if !fileExists(dir) {
+		err := os.MkdirAll(dir, 0755)
+		if err != nil {
+			return fmt.Errorf("%w making directory: %s", err, dir)
+		}
+	}
+
+	out, err := os.Create(destination)
+	if err != nil {
+		return fmt.Errorf("%w creating %s", err, destination)
+	}
+	defer out.Close()
+
+	zipW := zip.NewWriter(out)
+	zipW.RegisterCompressor(zip.Deflate, func(out io.Writer) (io.WriteCloser, error) {
+		return flate.NewWriter(out, flate.DefaultCompression)
+	})
+	defer zipW.Close()
+
+	for _, source := range sources {
+		err := writeWalk(zipW, source, destination)
+		if err != nil {
+			return fmt.Errorf("%w walking %s", err, source)
+		}
+	}
+
+	return nil
+}
+
+// UnZip unpacks the .zip file at source to destination.
+func UnZip(source, destination string) error {
+	destinationDir := filepath.Dir(destination)
+
+	if !fileExists(destinationDir) {
+		err := os.MkdirAll(destinationDir, 0755)
+		if err != nil {
+			return fmt.Errorf("preparing destination: %v", err)
+		}
+	}
+
+	zipR, err := zip.OpenReader(source)
+	if err != nil {
+		return fmt.Errorf("opening source file: %v", err)
+	}
+	defer zipR.Close()
+
+	for _, f := range zipR.File {
+		err := extractAndWriteFile(f, destination)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func extractAndWriteFile(f *zip.File, destination string) error {
+	rc, err := f.Open()
+	if err != nil {
+		return fmt.Errorf("failed to open file: %v", err)
+	}
+	defer rc.Close()
+
+	path := filepath.Join(destination, f.Name)
+
+	// Check for ZipSlip (Directory traversal)
+	if !strings.HasPrefix(path, filepath.Clean(destination)+string(os.PathSeparator)) {
+		return fmt.Errorf("illegal file path: %s", path)
+	}
+
+	if f.Mode() == os.ModeSymlink {
+		// We skip symlink for security reasons
```

Comment:
I think we may need logged warnings for this (if possible)

Reply:
PS: I'm not sure how much of this is duplicated by the updater (using `tar`): https://github.com/DataDog/datadog-agent/blob/49f6e7ff8426ecc646e8ce1c9e452d149bffefdd/pkg/fleet/internal/tar/tar.go#L27 and can be joined (could put a question back on the table if we only use zips)

CC: @arbll since it could be interesting to combine reusable code

Reply:
That would nice indeed! It might be a bit tricky tho because our usecase:
- requires symlinks and hardlinks today
- considers the extracted archives as "trusted" as they are signed during updates and going to be executed shortly after